File: src/checker/propagation.ts
import type { Type } from "./types.js";
import type { TangleDiagnostic } from "../model.js";
import type { ErrorRegistry } from "./errors.js";

// checkPropagation: strips error variants from a sum type on ?
// Returns [okType, diagnostics]
export function checkPropagation(type: Type, errorRegistry: ErrorRegistry): [Type, TangleDiagnostic[]] {
  const diags: TangleDiagnostic[] = [];
  if (type.kind === "sum") {
    const oks: Type[] = [];
    for (const v of type.variants) {
      const name = getVariantName(v);
      if (name && errorRegistry.isError(name)) {
        // error variant — stripped
      } else {
        oks.push(v);
      }
    }
    if (oks.length === 0) {
      diags.push({
        code: "TANGLE_TYPE_ALL_ERROR",
        message: "All variants are errors — nothing to unwrap",
        span: { file: "", startLine: 1, startColumn: 1, endLine: 1, endColumn: 1 },
      });
      return [{ kind: "primitive", name: "Bool" }, diags];
    }
    if (oks.length === 1) return [oks[0]!, diags];
    // Multiple ok variants: return the sum of oks
    return [{ kind: "sum", variants: oks }, diags];
  }
  // Not a sum type — ? has no error to propagate
  return [type, diags];
}

function getVariantName(t: Type): string | null {
  if (t.kind === "struct") return t.name;
  if (t.kind === "primitive") return t.name;
  if (t.kind === "interface") return t.name;
  return null;
}

File: compiler/tangle-cli/src/checker/propagation.rs
use crate::checker::types::*;
use crate::checker::errors::ErrorRegistry;
use crate::model::TangleDiagnostic;

/// Strip error variants from a sum type when ? operator is used.
/// Returns the ok type(s) remaining after removing error variants.
pub fn check_propagation(ty: &Type, registry: &ErrorRegistry) -> (Type, Vec<TangleDiagnostic>) {
    let mut diags = vec![];
    match ty {
        Type::Sum(sum) => {
            let ok_variants: Vec<Type> = sum.variants.iter()
                .filter(|v| {
                    if let Type::Primitive(p) = v {
                        !registry.is_error(&p.name)
                    } else { true }
                })
                .cloned()
                .collect();

            if ok_variants.is_empty() {
                diags.push(TangleDiagnostic {
                    code: "TANGLE_TYPE_ALL_ERROR".into(),
                    message: "All variants are errors — propagation leaves no ok type".into(),
                    span: crate::model::SourceSpan {
                        file: String::new(), start_line: 0, start_column: 0, end_line: 0, end_column: 0,
                    },
                });
                (Type::Primitive(PrimitiveType { name: "Bool".into() }), diags)
            } else if ok_variants.len() == 1 {
                (ok_variants[0].clone(), diags)
            } else {
                (Type::Sum(SumType { variants: ok_variants }), diags)
            }
        }
        _ => (ty.clone(), diags),
    }
}
